cmd/srcd/cmd/parse_test.go

```diff
@@ -0,0 +1,257 @@
+		filename: "hello.php",
+		lang:     "php",
+	},
+}
+
+// The suite downloads all the test cases files only once, on the suite setup
+func (s *ParseTestSuite) SetupSuite() {
+	var err error
+	s.testDir, err = ioutil.TempDir("", "parse-test")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Download a hello world file for each supported language
+	for _, tc := range testCases {
+		path := filepath.Join(s.testDir, tc.filename)
+		s.downloadFile(tc.url, path)
+	}
+}
+
+func (s *ParseTestSuite) TearDownSuite() {
+	os.RemoveAll(s.testDir)
+}
+
+func (s *ParseTestSuite) SetupTest() {
+
+}
+
+func (s *ParseTestSuite) TearDownTest() {
+	s.RunStop(context.Background())
+}
+
+func (s *ParseTestSuite) downloadFile(url, path string) {
+	s.T().Helper()
+
+	out, err := os.Create(path)
+	s.Require().NoError(err)
+	defer out.Close()
+
+	resp, err := http.Get(url)
+	s.Require().NoError(err)
+	defer resp.Body.Close()
+
+	_, err = io.Copy(out, resp.Body)
+	s.Require().NoError(err)
+}
+
+func (s *ParseTestSuite) TestDriversList() {
+	require := s.Require()
+
+	out, err := s.RunCommand(context.TODO(), "parse", "drivers", "list")
+	outStr := out.String()
+
+	require.NoError(err, outStr)
+
+	/* Example output:
+
+	LANGUAGE	VERSION
+	----------	----------
+	python		v2.8.0
+	cpp		v1.1.0
+	java		v2.5.0
+	javascript	v2.6.0
+	bash		v2.4.0
+	ruby		v2.9.0
+	go		v2.5.0
+	csharp		v1.4.0
+	php		v2.7.0
+	*/
+
+	// Simple checks to see if it's the table, and contains a known driver
+	require.Contains(outStr, "LANGUAGE	VERSION")
+	expected := regexp.MustCompile(`javascript	v\S+`)
+	require.Regexp(expected, outStr)
+}
+
+func (s *ParseTestSuite) TestLang() {
+	for _, tc := range testCases {
+		s.T().Run(tc.filename, func(t *testing.T) {
+			require := require.New(t)
+
+			path := filepath.Join(s.testDir, tc.filename)
+
+			// Check the language is detected
+			out, err := s.RunCommand(context.TODO(), "parse", "lang", path)
+			require.NoError(err, out.String())
+			require.Equal(tc.lang+"\n", out.String())
+		})
+	}
+}
+
+// same as RunCommand, but captures only stdout instead of stdout + stderr
+func (s *ParseTestSuite) runCommandStdout(ctx context.Context, cmd string, args ...string) (*bytes.Buffer, error) {
+	var out bytes.Buffer
+
+	command := s.CommandContext(ctx, cmd, args...)
+	command.Stdout = &out
+
+	return &out, command.Run()
+}
+
+func (s *ParseTestSuite) TestUast() {
+	modes := []string{"default-mode", "semantic", "annotated", "native"}
+	langOpts := []string{"no-lang-option", "--lang"}
+	queryOpts := []string{"no-query-option", "--query"}
+
+	for _, tc := range testCases {
+		s.T().Run(tc.filename, func(t *testing.T) {
+			path := filepath.Join(s.testDir, tc.filename)
+
+			for _, mode := range modes {
+				t.Run(mode, func(t *testing.T) {
+
+					for _, lang := range langOpts {
+						t.Run(lang, func(t *testing.T) {
+
+							for _, query := range queryOpts {
+								t.Run(query, func(t *testing.T) {
+
+									require := require.New(t)
+
```

Comment:
wow. The cyclomatic complexity of this function must be crazy.

Reply:
And the code is super ugly, the level of indentation alone is an eye sore. But... it's for the tests, and only to mix all the table test cases.
Should we try to improve this somehow?

Reply:
to be clear: I propose to rewrite it to avoid so many inner for loops. It's hard to read.

Reply:
without removing for loops themselves but at least separating it from actual tests we can do something like:
```go
func Test() {
  expandedTestCases := []expandedTestCase{}
  for _, tc := range testCases {
	for _, mode := range modes {
		for _, lang := range langOpts {
			for _, query := range queryOpts {
				name := ""
				expandedTestCases = append(expandedTestCases, expandedTestCase{...})
			}
		}
	}
  }

  for _ tc := range expandedTestCases {
	  // actual test
  }
}

type expandedTestCase struct {
	testCase

	name string
	mode string
	lang string
	query string
}
```

Reply:
another option is keep this for loops with t.Run as is but move "meat" to separate function.

Reply:
I gave a try in 41aadd2, I see I found a solution similar to your first proposal.

Reply:
I agree that readability is important.
I'd purpose a cleaner way to combine the arguments (permutations is about ordering, a different thing)
https://play.golang.org/p/i991wbz9WaS
imo, that way is clear that you're combining queries, modes, and langs, per test case.

I'd also extract the generation of the command arguments from the `TestUast` test case

Reply:
ok, done in 5b68fd3